contribs/gnodev/pkg/browser/model.go

```diff
@@ -0,0 +1,572 @@
+	Renderer        *lipgloss.Renderer
+	Readonly        bool
+	Banner          ModelBanner
+}
+
+const DefaultGnoLandPrefix = "gno.land/"
+
+func DefaultConfig() Config {
+	return Config{
+		Logger:          log.NewNoopLogger(),
+		URLPrefix:       "gno.land/",
+		Renderer:        lipgloss.DefaultRenderer(),
+		URLDefaultValue: "gnoland/home",
+	}
+}
+
+type model struct {
+	render *lipgloss.Renderer
+	client *NodeClient
+	logger *slog.Logger
+
+	// misc
+	banner          ModelBanner
+	bannerDiscarded bool
+
+	// Viewport
+	zone           *zone.Manager
+	ready          bool
+	viewport       viewport.Model
+	height, width  int
+	readonly       bool
+	messageDisplay bool
+
+	// Nav
+	taskLoader LoaderModel
+
+	pageurls map[string]string
+	history  *clist.List
+	current  *clist.Element
+
+	// Url
+	urlInput  textinput.Model
+	urlPrefix string
+
+	// Commands
+	listFuncs    FuncListModel
+	commandInput textinput.Model
+	commandFocus bool
+}
+
+func initURLInput(prefix string, r *lipgloss.Renderer) textinput.Model {
+	ti := textinput.New()
+	ti.Placeholder = "r/gnoland/blog" // XXX: Use as example, customize this ?
+	ti.Focus()
+	ti.CharLimit = 156
+	ti.PromptStyle = promptStyle(r)
+	ti.Prompt = prefix + "/"
+
+	return ti
+}
+
+func initCommandInput(r *lipgloss.Renderer) textinput.Model {
+	ti := textinput.New()
+	ti.Placeholder = ""
+	ti.CharLimit = 156
+	ti.PromptStyle = promptStyle(r)
+	ti.Prompt = "> "
+
+	return ti
+}
+
+func New(cfg Config, client *gnoclient.Client) tea.Model {
+	renderer := lipgloss.DefaultRenderer()
+	if cfg.Renderer != nil {
+		renderer = cfg.Renderer
+	}
+
+	// Setup url input
+	urlinput := initURLInput(cfg.URLPrefix, renderer)
+
+	path := cleanupRealmPath(cfg.URLPrefix, cfg.URLDefaultValue)
+	urlinput.SetValue(path)
+
+	// Setup cmd input
+	cmdinput := initCommandInput(renderer)
+
+	// XXX: Customize this
+	base := gnoclient.BaseTxCfg{
+		GasFee:    "1000000ugnot",
+		GasWanted: 2000000,
+	}
+
+	nodeclient := NewNodeClient(cfg.Logger, base, client)
+	return &model{
+		logger:     cfg.Logger,
+		render:     cfg.Renderer,
+		readonly:   cfg.Readonly,
+		client:     nodeclient,
+		taskLoader: newLoaderModel(),
+
+		banner:          cfg.Banner,
+		bannerDiscarded: cfg.Banner.Empty(),
+
+		urlInput:  urlinput,
+		urlPrefix: cfg.URLPrefix,
+
+		commandInput: cmdinput,
+		listFuncs:    newFuncList(),
+
+		zone:     zone.New(),
+		pageurls: map[string]string{},
+		history:  clist.New(),
+	}
+}
+
+func (m model) Init() tea.Cmd {
+	m.history.Init()
+	return m.banner.Init()
+}
+
+type fetchRealmMsg struct {
+	realmPath string
+}
+
+func FetchRealm(path string) tea.Cmd {
+	return func() tea.Msg { return fetchRealmMsg{path} }
+}
+
+func RefreshRealm() tea.Cmd {
+	return func() tea.Msg { return fetchRealmMsg{""} }
+}
+
+type renderUpdateMsg struct {
+	Render []byte
+	Funcs  vm.FunctionSignatures
+	Error  error
+}
+
+func (m *model) RenderUpdate(path string) tea.Cmd {
+	return func() tea.Msg {
+		var msg renderUpdateMsg
```

Comment:
My theory on this is that you should only do that if you have more than 2 vars, because:

```go
var x Y
var z W
```

is still shorter than this with the same readability

```go
var (
	x Y
	z W
)
```

Reply:
Well, it will always be shorter. I just mean that for two variables, it seems a bit overwhelming.